compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/FunctionalConstructorBuilder.java

```diff
@@ -0,0 +1,93 @@
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.wso2.ballerinalang.compiler.util;
+
+
+import org.ballerinalang.model.symbols.SymbolKind;
+import org.wso2.ballerinalang.compiler.semantics.model.Scope;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BConstructorSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BInvokableTypeSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BPackageSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BVarSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.SymTag;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BInvokableType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Build functional constructor symbol.
+ *
+ * @since 1.3.0
+ */
+public class FunctionalConstructorBuilder {
+    private ArrayList<BVarSymbol> params;
+    private String name;
+    private BPackageSymbol langlibPkg;
+    private BType constructedType;
+
+    private FunctionalConstructorBuilder(String name, BPackageSymbol langlibPkg, BType constructedType) {
+        this.name = name;
+        this.langlibPkg = langlibPkg;
+        this.constructedType = constructedType;
+        this.params = new ArrayList<>();
+    }
+
+    public static FunctionalConstructorBuilder newConstructor(String name,
+                                                              BPackageSymbol langlibModule,
+                                                              BType constructedType) {
+        return new FunctionalConstructorBuilder(name, langlibModule, constructedType);
+    }
+
+    private FunctionalConstructorBuilder addParam(String name, BType type, boolean isDefaultable) {
+        BVarSymbol paramSymbol = new BVarSymbol(0, new Name(name), langlibPkg.pkgID, type, null);
+        paramSymbol.defaultableParam = isDefaultable;
+        params.add(paramSymbol);
+        return this;
+    }
+
+    public FunctionalConstructorBuilder addParam(String name, BType type) {
+        return addParam(name, type, false);
+    }
+
+    public FunctionalConstructorBuilder addDefaultableParam(String name, BType type) {
+        return addParam(name, type, true);
+    }
+
+    public BConstructorSymbol build() {
+        List<BType> paramTypes = new ArrayList<>();
```

Comment:
Think we can provide initial capacity for array list size for `new ArrayList<>(params.size())`

Reply:
Done